compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1429,6 +1429,71 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         true
     }
 
+    // Suggest to change `Option<&Vec<T>>::unwrap_or(&[])` to `Option::map_or(&[], |v| v)`.
+    #[instrument(level = "trace", skip(self, err, provided_expr))]
+    pub(crate) fn suggest_deref_unwrap_or(
+        &self,
+        err: &mut Diag<'_>,
+        error_span: Span,
+        callee_ty: Option<Ty<'tcx>>,
+        call_ident: Option<Ident>,
+        expected_ty: Ty<'tcx>,
+        provided_ty: Ty<'tcx>,
+        provided_expr: &Expr<'tcx>,
+        is_method: bool,
+    ) {
+        if !is_method {
+            return;
+        }
+        let Some(callee_ty) = callee_ty else {
+            return;
+        };
+        let ty::Adt(callee_adt, _) = callee_ty.peel_refs().kind() else {
+            return;
+        };
+        if !self.tcx.is_diagnostic_item(sym::Option, callee_adt.did()) {
+            return;
+        }
+
+        if call_ident.map_or(true, |ident| ident.name != sym::unwrap_or) {
+            return;
+        }
+
+        let ty::Ref(_, peeled, _mutability) = provided_ty.kind() else {
+            return;
+        };
+
+        // NOTE: Can we reuse `suggest_deref_or_ref`?
+
+        // Create an dummy type `&[_]` so that both &[] and `&Vec<T>` can coerce to it.
+        let dummy_ty = if let ty::Array(elem_ty, size) = peeled.kind()
+            && let ty::Infer(_) = elem_ty.kind()
+            && size.try_eval_target_usize(self.tcx, self.param_env) == Some(0)
+        {
+            let slice = Ty::new_slice(self.tcx, *elem_ty);
+            Ty::new_imm_ref(self.tcx, self.tcx.lifetimes.re_static, slice)
+        } else {
+            provided_ty
+        };
+
+        if !self.can_coerce(expected_ty, dummy_ty) {
+            return;
+        }
+        let (provided_snip, applicability) =
+            match self.tcx.sess.source_map().span_to_snippet(provided_expr.span) {
+                Ok(snip) => (snip, Applicability::MachineApplicable),
+                Err(_) => ("/* _ */".to_owned(), Applicability::MaybeIncorrect),
+            };
+        let sugg = &format!("map_or({provided_snip}, |v| v)");
+        err.span_suggestion_verbose(
+            error_span,
+            "use `Option::map_or` to deref inner value of `Option`",
+            sugg,
+            applicability,
+        );
```

Comment:
Instead of `span_suggestion_verbose` we can use `multipart_suggestion_verbose` for this, as there are effectively multiple things to change. You can take the `call_ident.span` to replace it with `map_or` and then `provided_expr.span.shrink_to_hi()`:

```suggestion
        err.multipart_suggestion_verbose(
            "use `Option::map_or` to deref the inner value of the `Option`",
            vec![
                (call_ident.span, "map_or".to_string()),
                (provided_expr.span.shrink_to_hi(), "|v| v".to_string()),
            ],
            applicability,
        );
```